ecommerce/extensions/fulfillment/modules.py

```
@@ -507,3 +509,129 @@ def send_email(self, order):
+                self._post_to_entitlement_api(data, site=order.site)
+
+                line.set_status(LINE.COMPLETE)
+
+                audit_log(
+                    'line_fulfilled',
+                    order_line_id=line.id,
+                    order_number=order.number,
+                    product_class=line.product.get_product_class().name,
+                    course_id=course_key,
+                    mode=mode,
+                    user_id=order.user.id,
+                )
+            except (ConnectionError, Timeout):
+                logger.error(
+                    'Unable to fulfill line [%d] of order [%s] due to a network problem', line.id, order.number
+                )
+                line.set_status(LINE.FULFILLMENT_NETWORK_ERROR)
+
+        logger.info('Finished fulfilling "Course Entitlement" product types for order [%s]', order.number)
+        return order, lines
+
+    def revoke_line(self, line):
+        try:
+            logger.info('Attempting to revoke fulfillment of Line [%d]...', line.id)
+
+            mode = mode_for_seat(line.product)
+            course_key = line.product.attr.course_key
+            data = {
+                'user': line.order.user.username,
+                'is_active': False,
+                'mode': mode,
+                'course_entitlement_details': {
+                    'course_id': course_key,
+                },
+            }
+
+            self._post_to_entitlement_api(data, site=line.order.site)
+
+            audit_log(
+                'line_revoked',
+                order_line_id=line.id,
+                order_number=line.order.number,
+                product_class=line.product.get_product_class().name,
+                course_id=course_key,
+                certificate_type=getattr(line.product.attr, 'certificate_type', ''),
+                user_id=line.order.user.id
+            )
+
+            return True
+        except Exception:  # pylint: disable=broad-except
```

What kind of exception(s) can we expect to be raised here? Should they possible be handled / logged differently?

---

Did you get a chance to look at this one @mjfrey ?

---

We generally catch all Exceptions here rather than enumerating all possible exceptions.  Take a look at the Seat revoke.  If anything goes wrong, we just do  not revoke.

---

👍 works for me